GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs

```diff
@@ -307,37 +313,32 @@ private void Update()
                 return;
             }
 
-            if (Environment.TickCount < _nextUpdateTime)
+            if (_nextUpdateTime - Environment.TickCount > 0)
             {
                 return;
             }
 
             // don't update status while repository is being modified by GitExt,
-            // repository status will change after these actions.
+            // repository status may change after these actions.
             if (UICommandsSource.UICommands.RepoChangedNotifier.IsLocked ||
                 (GitVersion.Current.RaceConditionWhenGitStatusIsUpdatingIndex && Module.IsRunningGitProcess()))
             {
-                ScheduleNextUpdateTime(0);
                 return;
             }
 
             lock (_statusSequence)
             {
-                // If the previous status call hasn't exited yet,
-                // schedule new update when command is finished
                 if (_commandIsRunning)
                 {
-                    ScheduleNextUpdateTime(0);
                     return;
                 }
 
                 _commandIsRunning = true;
-            }
-
-            _nextIsInteractive = false;
+                _pendingUpdate = false;
 
-            // Schedule update every 5 min, even if we don't know that anything changed
-            ScheduleNextUpdateTime(PeriodicUpdateInterval);
+                // Schedule update every 5 min, even if we don't know that anything changed
+                _nextUpdateTime = Environment.TickCount + PeriodicUpdateInterval;
+            }
 
             if (!Directory.Exists(_workTreeWatcher.Path))
             {
```

Comment:
`_commandIsRunning = false` should be set (under `lock (_statusSequence)`) before exiting this function here, shouldn't it?
This should also be done if the `RunAsync` fails or is not reached due to an exception.

Reply:
If there is a problem, the handling is disabled, now and earlier.
You will have to restart by open a new repo.
Previously the handling was deactivated but status variables were reset.

Reply:
I've seen too many race conditions.
This one might be synthetic and is perhaps resolved at another place (where?) - at least by reloading the repo. (Though I consider the latter to be a only workaround.)

```C#
            lock (_statusSequence)
            {
                if (_commandIsRunning)
                {
                    return;
                }

                _commandIsRunning = true;
                _pendingUpdate = false;

                // Schedule update every 5 min, even if we don't know that anything changed
                _nextUpdateTime = Environment.TickCount + PeriodicUpdateInterval;
            }

            if (!Directory.Exists(_workTreeWatcher.Path))
            {
                // The directory no longer exists, watcher cannot be enabled
                return;
            }

            // snip

            ThreadHelper.JoinableTaskFactory.RunAsync(
                    async () =>
                    {
                         // snip

                         _commandIsRunning = false;
                    }).FileAndForget();
```
If the directory is unavailable temporarily, `_commandIsRunning` will never be reset to `false` unless the state `GitStatusMonitorState.Running` is re-entered, because this function will return before it will start the command. In the next calls of this function, it will exit early regardless whether the directory is available again.
Exceptions from e.g. `Directory.Exists()`, `_statusSequence.Next()` or `JoinableTaskFactory.RunAsync` could have the same effect.

Reply:
!Directory.Exists(_workTreeWatcher.Path) is related to 
_workTreeWatcher.EnableRaisingEvents = true;
This should be moved into lock() to handle this situation which is relevant if you have the repo on a network drive or USB stick

The exceptions need to open a new repo.
I would like to handle that in a forced refresh too, but this is currently a part of a normal flow, so it will require a rewrite, like:
https://github.com/gerhardol/gitextensions/blob/bugfix/i7183-gitstatus-alternative/GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs#L376
I prefer the current

Reply:
Both variants do not address the case where the async task is not started at all because of an exception before `RunAsync`.

Reply:
> Both variants do not address the case where the async task is not started at all because of an exception before RunAsync.

That would require handling _commandIsRunning just before Async
Do not like that either, pushed an example to the alt branch

To be clear: This is a very relevant discussion, even if I not want to change.

Reply:
The alt branch looks much safer. But it must be checked that `CancellationTokenSequence.Next()` cannot cause a deadlock if called under `lock`.
> even if I not want to change

What are your concerns?

Reply:
> What are your concerns?

My concern is that this adds complexity for something that will not improve the functionality anyway.
I could ask for another opinion, but that will likely be "do the complete implementation", so I will clean the alt branch up.

Reply:
Pushed changes
Will squash when approved